tools/mpremote: Add readline support to mount.

andrewleech opened the pull request:
### Summary

This significantly speeds up readline on files opened directly from a mpremote mount.

The virtual file object created by mpremote on the device implemts readline locally (on device) by reading single byte by byte from the host computer which is quite slow.

This PR adds a new command to the mount subsystem to forward any readline calls directly to the host.

### Testing

This was tested when first written about a year ago, I haven't tested it again yet.

### Trade-offs and Alternatives

This make the in-ram RemoteFile class instance slightly larger though I expect the different is rather insignificant. This class is only used ephemerally while mount is running though.

Josverl commented:
Thanks, I've needed to work around this slowness/instability , glad you created a fix.

dpgeorge commented on tools/mpremote/mpremote/transport_serial.py:
```diff
@@ -382,13 +382,14 @@ def umount_local(self):
-    "CMD_MKDIR": 11,
-    "CMD_RMDIR": 12,
+    "CMD_READLINE": 6,
+    "CMD_READ": 7,
```
please swap READ and READLINE (so READ is first), to match the use below

andrewleech replied:
Thanks for the review @dpgeorge I made all the changes suggested and ran into a weird bug:
```
>>> anl@STEP:~/micropython$ mpremote mount .
Traceback (most recent call last):
  File "/home/anl/.local/bin/mpremote", line 8, in <module>
    sys.exit(main())
  File "/home/anl/micropython/tools/mpremote/mpremote/main.py", line 538, in main
    handler_func(state, args)
  File "/home/anl/micropython/tools/mpremote/mpremote/commands.py", line 440, in do_mount
    state.transport.mount_local(path, unsafe_links=args.unsafe_links)
  File "/home/anl/micropython/tools/mpremote/mpremote/transport_serial.py", line 302, in mount_local
    self.exec(fs_hook_code)
  File "/home/anl/micropython/tools/mpremote/mpremote/transport_serial.py", line 291, in exec
    raise TransportExecError(ret, ret_err.decode())
mpremote.transport.TransportExecError: Traceback (most recent call last):
  File "<stdin>", line 145
SyntaxError: invalid syntax for number
```

Took me a while to remember that the order here was a neccesary workaround for:
``` python
# Apply basic compression on hook code.
for key, value in fs_hook_cmds.items():
    fs_hook_code = re.sub(key, str(value), fs_hook_code)
```
If the command replacement loop handles `CMD_READ` first then it replaces the usage of `CMD_READLINE` with `6LINE` in the hook code. A comment, or a fixed replacement loop, would have been helpful for us both!

dpgeorge commented on tools/mpremote/mpremote/transport_serial.py:
```diff
@@ -571,12 +572,14 @@ def readinto(self, buf):
+        buf = bytearray(n)
+        c.rd_bytes(buf)
+        c.end()
+        return bytes(buf)
```
This needs to return a str or bytes.  And `rd_bytes` automatically reads the length for you.  I suggest following how read is implemented:
```py
        c = self.cmd
        c.begin(CMD_READLINE)
        c.wr_s8(self.fd)
        data = c.rd_bytes(None)
        c.end()
        if self.is_text:
            data = str(data, 'utf8')
        else:
            data = bytes(data)
        return data
```

dpgeorge commented on tools/mpremote/mpremote/transport_serial.py:
```diff
@@ -866,6 +869,15 @@ def do_read(self):
+    def do_readline(self):
+        fd = self.rd_s8()
+        buf = self.data_files[fd][0].readline()
+        self.wr_u32(len(buf))
```
This `wr_u32` is not needed, `wr_bytes` will do it for you.

pi-anl committed:
tools/mpremote: Optimise readline support in mount.

This significantly speeds up readline on files opened directly from an
mpremote mount.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

andrewleech commented:
Thanks @dpgeorge I've implemented the suggestions and retested with an esp32s3

**Before:**
```
anl@STEP:~/micropython$ mpremote mount .
Local directory . is mounted at /remote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] or Ctrl-x to exit this shell
>
MicroPython v1.24.0-1.gb745173f3f.dirty on 2024-10-27; Generic ESP32S3 module with ESP32S3
Type "help()" for more information.
>>>
paste mode; Ctrl-C to cancel, Ctrl-D to finish
=== import time
=== f = open("ACKNOWLEDGEMENTS", "r")
=== t0 = time.ticks_ms()
=== count = 0
=== for line in f.readline():
===     count += 1
=== t1 = time.ticks_ms()
=== taken = time.ticks_diff(t1, t0)
=== print(f"taken {taken}ms for {count} lines")
===
taken 187ms for 65 lines
>>>
```

**After:**
```
anl@STEP:~/micropython$ mpremote mount .
Local directory . is mounted at /remote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] or Ctrl-x to exit this shell
>
MicroPython v1.24.0-1.gb745173f3f.dirty on 2024-10-27; Generic ESP32S3 module with ESP32S3
Type "help()" for more information.
>>>
paste mode; Ctrl-C to cancel, Ctrl-D to finish
=== import time
=== f = open("ACKNOWLEDGEMENTS", "r")
=== t0 = time.ticks_ms()
=== count = 0
=== for line in f.readline():
===     count += 1
=== t1 = time.ticks_ms()
=== taken = time.ticks_diff(t1, t0)
=== print(f"taken {taken}ms for {count} lines")
===
taken 8ms for 65 lines
>>>
```

dpgeorge commented on tools/mpremote/mpremote/transport_serial.py:
```diff
@@ -746,15 +751,18 @@ def __mount():
+# Replace hook command names with the integer command index.
+# Handle keys in reverse sort order to to avoid eg. READ replacing half of READLINE
+for key, value in reversed(sorted(fs_hook_cmds.items())):
+    fs_hook_code = re.sub(key, str(value), fs_hook_code)
```
Maybe it's better to make a more sophisticated regular expression that matches stand and end of word, eg `r"\W" + key + r"\W"`.

andrewleech replied:
Yeah I had thought that briefly... it would be more explicit to fix the regex

dpgeorge commented on tools/mpremote/mpremote/transport_serial.py:
```diff
@@ -747,7 +752,7 @@ def __mount():
 # Apply basic compression on hook code.
 for key, value in fs_hook_cmds.items():
-    fs_hook_code = re.sub(key, str(value), fs_hook_code)
+    fs_hook_code = re.sub(rf'(\W){key}(\W)', rf"\g<1>{value}\g<2>", fs_hook_code)
```
How about replacing the entire loop with a single line:
```py
fs_hook_code = re.sub(r"CMD_[A-Z]+", lambda match: fs_hook_cmds[match.group(0)], fs_hook_code)
```

andrewleech replied:
I like it. Minor tweaks to make it work and it tests out fine with the script I used earlier. Pushed now.

dpgeorge replied:
Very good!

dpgeorge committed:
tools/mpremote/tests: Add test for RemoteFile.readline.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Thanks, this looks really good now.

I've added a small test for `RemoteFile.readline`.  See what you think.

andrewleech commented:
> I've added a small test for `RemoteFile.readline`.  See what you think.

I must confess I didn't realise there were tests for mpremote, thanks for adding this. It looks clean and simple, certainly great to have the test coverage.
